eng/scripts/automation/Invoke-GenerateAndBuild.ps1

```diff
@@ -41,27 +42,59 @@ if ($autorestConfig -ne "") {
+$generatedSDKPackages = New-Object 'Collections.Generic.List[System.Object]'
 
 # $service, $serviceType = Get-ResourceProviderFromReadme $readmeFile
 $sdkPath =  (Join-Path $PSScriptRoot .. .. ..)
 $sdkPath = Resolve-Path $sdkPath
 $sdkPath = $sdkPath -replace "\\", "/"
 
-$readme = ""
-if ($commitid -ne "") {
-  if ($repoHttpsUrl -ne "") {
-    $readme = "$repoHttpsUrl/blob/$commitid/$readmeFile"
+if ($readmeFile) {
+  Write-Host "swaggerDir:$swaggerDir, readmeFile:$readmeFile"
+
+  $readme = ""
+  if ($commitid -ne "") {
+    if ($repoHttpsUrl -ne "") {
+      $readme = "$repoHttpsUrl/blob/$commitid/$readmeFile"
+    } else {
+      $readme = "https://github.com/$org/azure-rest-api-specs/blob/$commitid/$readmeFile"
+    }
   } else {
-    $readme = "https://github.com/$org/azure-rest-api-specs/blob/$commitid/$readmeFile"
+    $readme = (Join-Path $swaggerDir $readmeFile)
   }
-} else {
-  $readme = (Join-Path $swaggerDir $readmeFile)
+  Invoke-GenerateAndBuildSDK -readmeAbsolutePath $readme -sdkRootPath $sdkPath -autorestConfigYaml "$autorestConfigYaml" -downloadUrlPrefix "$downloadUrlPrefix" -generatedSDKPackages $generatedSDKPackages
 }
 
-$generatedSDKPackages = New-Object 'Collections.Generic.List[System.Object]'
-Invoke-GenerateAndBuildSDK -readmeAbsolutePath $readme -sdkRootPath $sdkPath -autorestConfigYaml "$autorestConfigYaml" -downloadUrlPrefix "$downloadUrlPrefix" -generatedSDKPackages $generatedSDKPackages
+if ($relatedCadlProjectFolder) {
+  $caldFolder = (Join-Path $swaggerDir $relatedCadlProjectFolder) -replace "\\", "/"
+  $newPackageOutput = "newPackageOutput.json"
 
+  Push-Location $caldFolder
+  $cadlProjectYaml = Get-Content -Path "$caldFolder/cadl-project.yaml" -Raw
+
+  Install-ModuleIfNotInstalled "powershell-yaml" "0.4.1" | Import-Module
+  $yml = ConvertFrom-YAML $cadlProjectYaml
+  $sdkFolder = $yml["emitters"]["@azure-tools/cadl-csharp"]["sdk-folder"]
+  $projectFolder = (Join-Path $sdkPath $sdkFolder)
+  $projectFolder = $projectFolder -replace "\\", "/"
+  if ($projectFolder) {
+      $directories = $projectFolder.Split("/");
+      $count = $directories.Count
+      $projectFolder = $directories[0 .. ($count-2)] -join "/"
+      $service = $directories[-3];
+      $namespace = $directories[-2];
+  }
+  New-CADLPackageFolder -service $service -namespace $namespace -sdkPath $sdkPath -cadlInput $caldFolder/main.cadl -outputJsonFile $newpackageoutput
+  $newPackageOutputJson = Get-Content $newPackageOutput | Out-String | ConvertFrom-Json
+  $relativeSdkPath = $newPackageOutputJson.path
+  npm install
+  npx cadl compile --output-path $sdkPath --emit @azure-tools/cadl-csharp .
+  if ( !$?) {
+      Throw "Failed to generate sdk for cadl. exit code: $?"
+  }
+  GeneratePackage -projectFolder $projectFolder -sdkRootPath $sdkPath -path $relativeSdkPath -downloadUrlPrefix "$downloadUrlPrefix" -skipGenerate -generatedSDKPackages $generatedSDKPackages
+  Pop-Location
```

Comment:
Another instance where you should either eliminate this pattern, or put `trap { pop-location }` *immediately* after `push-location`.